validate.go

```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
 	return crawlers
 }()
 
-var regexps = func() []*regexp.Regexp {
-	regexps := make([]*regexp.Regexp, len(Crawlers))
+var pattern2literals = map[string][]string{
+	`[wW]get`:               {`wget`, `Wget`},
+	`Ahrefs(Bot|SiteAudit)`: {`AhrefsBot`, `AhrefsSiteAudit`},
+	`S[eE][mM]rushBot`:      {`SemrushBot`, `SeMrushBot`, `SEmrushBot`, `SEMrushBot`},
+	`Livelap[bB]ot`:         {`Livelapbot`, `LivelapBot`},
+	`[pP]ingdom`:            {`pingdom`, `Pingdom`},
+	`Bark[rR]owler`:         {`Barkrowler`, `BarkRowler`},
+	`^Apache-HttpClient`:    {`^Apache-HttpClient`},
+	`^LCC `:                 {`^LCC `},
+	`(^| )sentry\/`:         {`^sentry/`, ` sentry/`},
+	`^curl`:                 {`^curl`},
+	`[Cc]urebot`:            {`Curebot`, `curebot`},
+	`^PHP-Curl-Class`:       {`^PHP-Curl-Class`},
+	`(^| )PTST\/`:           {`^PTST/`, ` PTST/`},
+	`^BW\/`:                 {`^BW/`},
+}
+
+var pattern2mainLiteral = map[string]string{
+	`AdsBot-Google([^-]|$)`:             `AdsBot-Google`,
+	`BlogTraffic\/\d\.\d+ Feed-Fetcher`: `BlogTraffic/`,
+}
+
+func analyzePattern(pattern string) (olds []string, re *regexp.Regexp) {
+	literals, has := pattern2literals[pattern]
+	if has {
+		return literals, nil
+	}
+
+	re = regexp.MustCompile(pattern)
+	prefix, complete := re.LiteralPrefix()
+	if complete {
+		return []string{prefix}, nil
+	}
+
+	mainLiternal, has := pattern2mainLiteral[pattern]
+	if !has {
+		panic("don't know what to do with pattern: " + pattern)
+	}
+	return []string{mainLiternal}, re
+}
+
+type regexpPattern struct {
+	re    *regexp.Regexp
+	index int
+}
+
+type matcher struct {
+	replacer *strings.Replacer
+	regexps  []regexpPattern
+}
+
+var uniqueToken = hex.EncodeToString((&maphash.Hash{}).Sum(nil))
+
+const (
+	literalLabel = '-'
+	regexpLabel  = '*'
+)
+
+var m = func() matcher {
+	regexps := []regexpPattern{}
+	oldnew := make([]string, 0, len(Crawlers)*2)
+
+	// Put re-based patterns to the end to prevent AdsBot-Google from
+	// shadowing AdsBot-Google-Mobile.
+	var oldnew2 []string
+
 	for i, crawler := range Crawlers {
-		regexps[i] = regexp.MustCompile(crawler.Pattern)
+		literals, re := analyzePattern(crawler.Pattern)
+
+		label := literalLabel
+		num := i
+		if re != nil {
+			label = regexpLabel
+			num = len(regexps)
+			regexps = append(regexps, regexpPattern{
+				re:    re,
+				index: i,
+			})
+		}
+
+		replaceWith := fmt.Sprintf(" %s%c%05d ", uniqueToken, label, num)
+
+		for _, literal := range literals {
+			if re != nil {
+				oldnew2 = append(oldnew2, literal, replaceWith)
+			} else {
+				oldnew = append(oldnew, literal, replaceWith)
+			}
+		}
+	}
+	oldnew = append(oldnew, oldnew2...)
+
+	regexps2 := make([]regexpPattern, len(regexps))
```

Comment:
Save memory. `regexps` is allocated in `append` calls and it is likely to have extra capacity, because `append` reallocates with reserves in capacity. `regexps2` is allocated for exact size that is needed. We know the size only after finishing iterating over patterns, so we can not preallocate an array of the exact size in advance.

I added a comment in the code.